### PR TITLE
[MISC;ISSUE-12] Makefile APIs

### PR DESCRIPTION
**Pull request by deusebio**

### CODE REVIEW

**Review by nromeoarena — APPROVED** [review]
Ok for me

File: requirements/requirements_dev.in [hunk 1]
```diff
@@ -1,10 +1,12 @@
--r requirements.txt
--c requirements.txt
+-r requirements.in
+-c requirements.in
```
Comment (NicolaDonelli) [inline requirements/requirements_dev.in:2]: Why this?
We should restrict our development environment's requirement (requirements_dev) to exactly the same requirements as the production environment's ones (requirements.txt) and not to their open counterpart (requirements.in).

File: requirements/requirements_dev.in [hunk 2]
```diff
@@ -1,10 +1,12 @@
...
 pytest>=6.2.5
 pytest-cov>=3.0.0
-pip-tools>=6.4.0
-pre-commit>=2.16.0
+pip-tools==6.1.0
```
Comment (NicolaDonelli) [inline requirements/requirements_dev.in:5]: What is wrong with newer versions of pip-tools? Why limit this template to a specific (closed) version?

**Review by NicolaDonelli — CHANGES_REQUESTED** [review]
I do not like very much the changes to requirements files.
I am mildly against using closed dependencies in *.in files (both for dev and for prod requirements) since I feel this could become an unnecessary limit over time. On the other hand I am strongly against constraining `requirements_dev.in` to `requirements.in` instead to `requirements.txt` since constraining dev requirements to a list of open requirements (as `requirements.in` should be) would allow dev requirements to diverge from prod requirements also in the packages shared by these two environments (and this is not acceptable in my opinion).

Except for that, as suggested also in the issue, I'd also like a "run_ci" make command to execute locally actions in CI workflow. Do you think it could be possible to add?

File: requirements/requirements_dev.in (hunk 2 above)
Comment (deusebio) [reply requirements/requirements_dev.in:5]: Nothing, but I was just keeping this consistent with the version present in .pre-config-config.yaml. If we are not doing this, the two versions of pip-tools produce a slightly different requirements.txt and requirements_dev.txt files, and therefore this make pre-commit checks to NOT pass. If one wants to use pre-commit, the version of pip-tools MUST be the same between the environment and the pre-commit checks

**Comment by deusebio** [comment]
I did some modifications to the requirements files (including the upper bound) in order to get things running and produce consistent environments. I could not get the pipelines running without errors otherwise. What happened was in fact that the dev build would break when using "requirements.txt" and not bounding importlib-metadata. The thing is that conflicts with the requirements.txt versions would show up only when compiling the dev environment, but there the strict requirement would prevent to pick another version, hence the error. 

To be honest, this requirements issue is far from obvious and it should probably be tackled in a separated issue (this PR is only for writing the Makefile, not much about handling the requirements). I have updated the PR with a slightly revised version where

* we build the requirements.txt with the requirements_dev.in as constraints (and also importlib-metadata bound which is not taken otherwise since constraints seem NOT to follow transitive dependencies). 
* the requirements_dev.txt can be build using the requirements.txt AND requirements_dev.in together as inputs. 

These checks has nevertheless highlighted that Sphinx wants a importlib-metadata > 4.4 whereas flakes8 wants it to be <4.2. I have commented out Sphinx which is not used right now. But still, this requirements issue should be addressed in a separated issue, I believe. 

About the make run_ci, there is already a rule for doing so, which is the make checks. This runs mypy, linting and tests, which in turns depends on black and pip-tools for requirements, among other things. We could name it make ci if you prefer

**Comment by NicolaDonelli** [comment]
I still do not agree on constraining requirements files to open requirements files. As we discussed in person this could lead to having different versions of the same package in each environment and this is exactly what we want to avoid using pip-tools. 
Furthermore, I do not agree with constraining production requirements to dev requirements: it should be the other way around. 
For more details about this process see [pip-tools README](https://github.com/jazzband/pip-tools)

For what it takes the incompatibility between requirements of two of our chosen tools, I didn't go through the newly proposed configuration for requirements yet but, If the problem with requirements is this big, then we should face it before any other change that could impact or be impacted by them.
The fastest solution would be removing one of the two tools until we face this specific problem in another, dedicated, issue.

**Comment by nromeoarena** [comment]
> I still do not agree on constraining requirements files to open requirements files. As we discussed in person this could lead to having different versions of the same package in each environment and this is exactly what we want to avoid using pip-tools. Furthermore, I do not agree with constraining production requirements to dev requirements: it should be the other way around. For more details about this process see [pip-tools README](https://github.com/jazzband/pip-tools)
> 
> For what it takes the incompatibility between requirements of two of our chosen tools, I didn't go through the newly proposed configuration for requirements yet but, If the problem with requirements is this big, then we should face it before any other change that could impact or be impacted by them. The fastest solution would be removing one of the two tools until we face this specific problem in another, dedicated, issue.

Honestly I do agree with @NicolaDonelli especially on the fact that dev should be limited by prod requirements.
In any case it's preferable to have as strict as possible management of the requirements. So I think it's the case to investigate better the problem, as proposed, in another dedicated issue.

**Comment by deusebio** [comment]
So, shall we merge this PR as-is (which is really not quite about requirements) and move this to a different issue? Unless you have comments on makefiles and related...

**Comment by NicolaDonelli** [comment]
Agreed. 
I'll be fine with this PR when it will be reverted to the previous configuration of requirements.
So please revert to requirements_dev.in constrained to requirements.txt (removing the constraint added to requirements.txt to requirements_dev.in), eventually removing dependencies (like `flake8>=3.8.4`) that cause the conflict and their usages from Makefile (opening contextually an issue to highlight the problem).

**Review by NicolaDonelli — APPROVED** [review]
Still not a huge fan of having an upper bounded production requirement in a template but good to go.

**Pull request merged**